### PR TITLE
improvement: Use PcCollector for SemanticTokenProvider

### PR DESCRIPTION
**Pull request by jkciesluk**
Previously `SemanticTokenProvider` had its own tree traverser for collecting all symbol occurrences.
This PR changes it to use `PcCollector`, which should make optimisation and bugfixes easier. It should also simplify adding `SemanticTokens` for Scala 3

### CODE REVIEW

File: tests/slow/src/test/scala/tests/sbt/SbtBloopLspSuite.scala
Comment (jkciesluk) [inline tests/slow/src/test/scala/tests/sbt/SbtBloopLspSuite.scala:719]: It was incorrect previously, because of multiple symbols at the same position.

**Review by jkciesluk — COMMENTED** [review]
I've added some comments explaining changes

**Comment by jkciesluk** [comment]
Quick benchmarks show that after these changes, semantic tokens can work slower
After chages:
```
[info] Benchmark                                 (currentHighlightRequest)  Mode  Cnt    Score    Error  Units
[info] SemanticHighlightBench.semanticHighlight                    A.scala    ss   10    2.641 ±  0.508  ms/op
[info] SemanticHighlightBench.semanticHighlight        ClusterDaemon.scala    ss   10   49.994 ±  6.692  ms/op
[info] SemanticHighlightBench.semanticHighlight               Typers.scala    ss   10  161.479 ± 30.930  ms/op
[info] SemanticHighlightBench.semanticHighlight                Exprs.scala    ss   10   25.018 ± 14.654  ms/op
```
Before change:
```
[info] Benchmark                                 (currentHighlightRequest)  Mode  Cnt    Score    Error  Units
[info] SemanticHighlightBench.semanticHighlight                    A.scala    ss   10    1.912 ±  0.498  ms/op
[info] SemanticHighlightBench.semanticHighlight        ClusterDaemon.scala    ss   10   47.090 ± 16.405  ms/op
[info] SemanticHighlightBench.semanticHighlight               Typers.scala    ss   10  127.240 ± 45.553  ms/op
[info] SemanticHighlightBench.semanticHighlight                Exprs.scala    ss   10   22.164 ± 17.558  ms/op
```

On the other hand, they are more correct now and we can optimise `PcCollector`, which will improve also `highlight` and `rename`

File: tests/slow/src/test/scala/tests/sbt/SbtBloopLspSuite.scala
Comment (tgodzik) [reply tests/slow/src/test/scala/tests/sbt/SbtBloopLspSuite.scala:719]: Actually, I've noticed the bug just today, good catch!

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala
Comment (jkciesluk) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala:285]: This is copied from `soughtSymbols`. It typed trees we get `ident.symbol == NoSymbol`, so we use fallback method. It worked similarly in `SemanticTokenProvider` traverser

File: mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java
Comment (tgodzik) [reply mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java:44]: We can still keep the interface here, OffsetParams inherits from VirtualFileParams and we can match later on

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala:18]: If we change the parameter `VirtualFileParams` we can match here if it's an OffsetParams. If it's not OffsetParams we search for all symbols otherwise we use soughtSymbol?

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala:205]:
```suggestion
    params match {
     case params: OffsetParams => resultWithSought(params.offset)
     case  _ => resultAllOccurences().toList
    }
```
we could match here instead of changing the params to OffsetParams for semantic tokens

**Review by tgodzik — COMMENTED** [review]
Started reviewing and it looks good! I will take a further look later

File: mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala:38]: Should it always contain symbol? Do we use it when it's none?

File: tests/unit/src/test/resources/semanticTokens/example/NamedArguments.scala
Comment (tgodzik) [inline tests/unit/src/test/resources/semanticTokens/example/NamedArguments.scala:25]: This should be parameter, no?

**Review by tgodzik — CHANGES_REQUESTED** [review]
Actually, since most of the code is copied over I managed to finish it now :sweat_smile: 

Looks pretty awesome, I have a few comments, but nothing major.

File: mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala
Comment (jkciesluk) [reply mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala:38]: You're right. For now, there is None if symbol would be NoSymbol, but we can just keep the NoSymbol and skip the Option.

File: mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala
Comment (jkciesluk) [reply mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala]: It caused issues with anonymous classes, so I moved it to `pickFromTraversed` instead

File: mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala
Comment (jkciesluk) [inline mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala:230]: In case there are multiple symbols at the same position, we first filter out those with not matching name.
If there is still more than one symbol, we handle the case for class declaration
```scala
class Foo(a: Int, b: Int)
```
On `a` and `b` are two symbols: parameter in constructor method and `val`, we want to choose `val`.

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcCollector.scala:276]:
```suggestion
  def traverseSought(
      isSoughtSymbolTree: Tree => Boolean,
      filterBySoughtSymbols: (Set[Symbol] => Boolean) => Boolean
```

**Review by tgodzik — APPROVED** [review]
LGTM! If you could maybe rename the filter names a bit or add some comments that would be useful, but let's do it in a follow up so that the Scala 3 support is not blocked on this.

**Pull request merged**
